### PR TITLE
initial replacements of module system

### PR DESCRIPTION
**Pull request by endurance21**
This Pr is under the purview of   GSOC'20 project 
* all native `require and module.exports of nodejs` has been changed to `es6 export and import` in src folder !

### CODE REVIEW

**Comment by kyle1james** [comment]
See notes here https://pf-gsoc.slack.com/archives/C013F4XKRK4/p1589824958001200

**Comment by endurance21** [comment]
i have added p5.js file for the sole purpose of testing the APIs !

File: src/audiocontext.js [hunk 1]
```diff
@@ -2,121 +2,118 @@
...
-    } else if (elements instanceof Array && elements[0] instanceof p5.Element) {
-      elt = elements.map(function (e) {
-        return e.elt;
-      });
-    }
-    return StartAudioContext(audiocontext, elt, callback);
-  };
+// Tone and p5.sound share the same audio context
+Tone.context.dispose();
+Tone.setContext(audiocontext);
```
Comment (therewasaguy) [inline src/audiocontext.js:13]: I noticed some errors here. I think we need something like this to load the Tone Context module and to be able to set the context. For some reason we're not able to dispose of the context. We'll get an eslint error that Context is not used and we'll need to ignore that error or figure out how to use it :)

```
import Tone from 'Tone/core/Tone';
import Context from 'Tone/core/Context';

const audiocontext = new window.AudioContext();

// Tone and p5.sound share the same audio context
Tone.setContext(audiocontext);

File: src/app.js [hunk 2]
```diff
@@ -1,42 +1,36 @@
...
-  require('monosynth');
-  require('polysynth');
-  require('distortion');
-  require('audioVoice');
-  require('monosynth');
-  require('polysynth');
-
-  return p5SOUND;
-});
+import 'audioworklet-polyfill';
```
Comment (therewasaguy) [inline src/app.js:1]: for a future PR, this may be an opportunity to decouple p5.sound from p5. In the app module, we could check to see if p5 exists on the page. If it doesn't, we could behave differently, allowing p5.sound to be a stand-alone module. Just a thought...

File: src/audiocontext.js (hunk 1 above)
Comment (endurance21) [reply src/audiocontext.js:13]: yes that was really helpfull , the problem was due to Tone.Context was not in the scope of the current module ,
also 
doing 
import 'Tone/core/Context' solved the problem of linter!
@therewasaguy

**Comment by endurance21** [comment]
@therewasaguy  @kyle1james 
fixed the bug of setting the context of tone.js explicitly!
looking forward for your reviews !

File: package.json [hunk 3]
```diff
@@ -35,7 +35,7 @@
   "dependencies": {
     "audioworklet-polyfill": "^1.1.2",
     "startaudiocontext": "^1.2.1",
-    "tone": "0.10.0"
+    "tone": "^0.12.80"
```
[the remark this answers is not in the document]
Comment (endurance21) [reply package.json:38]: This was just an experiment oriented thing, like i was testing what things breaks if we upgarde 😵,
Anyways i dont see any diff after that, so i think we should switch back to 10th one

File: src/amplitude.js [hunk 4]
```diff
@@ -1,314 +1,311 @@
 'use strict';
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/amplitude.js:1]: I have a different view over this, actually our modules are transpiled into es5 or lower by babel,
So in browser our code would still be non es6 and hence browser need this
*use strict*
Declaration 
For operating in strict mode.

File: src/envelope.js [hunk 5]
```diff
@@ -1,909 +1,902 @@
...
-      this._rampAttackTime / this.checkExpInput(TCDenominator);
-    TCDenominator = Math.log(1.0 / this._rampLowPercentage);
-    this._rampDecayTC = this._rampDecayTime / this.checkExpInput(TCDenominator);
-  };
-
+import p5sound from './master';
+import Add from 'Tone/signal/Add';
+import Mult from 'Tone/signal/Multiply';
+import Scale from 'Tone/signal/Scale';
+import TimelineSignal from 'Tone/signal/TransportTimelineSignal.js';
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/envelope.js:7]: Actually i have to make a fix there, the actuall api is something different!
No worries, i will be doing that 
👍

File: src/polysynth.js [hunk 6]
```diff
@@ -1,481 +1,466 @@
...
+  //corresponds to the index of the audiovoice that is playing it
+  this.notes[note] = new TimelineSignal();
+  this.notes[note].setValueAtTime(currentVoice, acTime);
+
+  //Find the scheduled change in this._voicesInUse that will be previous to this new note
+  //Add 1 and schedule this value at time 't', when this note will start playing
+  var previousVal =
+    this._voicesInUse._searchBefore(acTime) === null
+      ? 0
+      : this._voicesInUse._searchBefore(acTime).value;
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/polysynth.js:304]: May be because this docs are for 13.8.25 version and we are using its
0.x version

File: src/audiocontext.js [hunk 7]
```diff
@@ -2,121 +2,118 @@
 
 global.TONE_SILENCE_VERSION_LOGGING = true;
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/audiocontext.js:3]: we are using the older version and hence there is no need to worry about that

File: src/polysynth.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (endurance21) [reply src/polysynth.js:304]: @therewasaguy  we reverted back to the older tone " 0.10.0"  now its fines i guess ?

File: src/polysynth.js [hunk 8]
```diff
@@ -1,481 +1,466 @@
...
+ * @for p5.PolySynth
+ * @param  {[type]} time  [description]
+ * @param  {[type]} value [description]
+ * @return {[type]}       [description]
+ */
+p5.PolySynth.prototype._updateAfter = function (time, value) {
+  if (this._voicesInUse._searchAfter(time) === null) {
+    return;
+  } else {
+    this._voicesInUse._searchAfter(time).value += value;
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/polysynth.js:337]: not a problem we reverted back to  the older tone !

File: src/polysynth.js [hunk 9]
```diff
@@ -1,481 +1,466 @@
...
-    //The audiovoice handles the actual scheduling of the note
-    if (typeof velocity === 'number') {
-      var maxRange = (1 / this._voicesInUse.getValueAtTime(acTime)) * 2;
-      velocity = velocity > maxRange ? maxRange : velocity;
+    //subtract 1 and schedule this value at time 't', when this note will stop playing
+    var previousVal = Math.max(~~this._voicesInUse.getValueAtTime(t).value, 1);
+    this._voicesInUse.setValueAtTime(previousVal - 1, t);
+    //Then update all scheduled values that follow to decrease by 1 but never go below 0
+    if (previousVal > 0) {
+      this._updateAfter(t, -1);
```
[the remark this answers is not in the document]
Comment (endurance21) [reply src/polysynth.js:413]: same ! here too

**Comment by endurance21** [comment]
@therewasaguy  i have made the requested changes and resolved the merge conflict too !
looking forward for the merge !

File: src/polysynth.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (therewasaguy) [reply src/polysynth.js:337]: we'll want to update at some point, but happy to do that in a separate PR

File: src/polysynth.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (therewasaguy) [reply src/polysynth.js:304]: yes, we'll want to update at some point, but we can do that in a separate PR if it's not required for this one

File: src/amplitude.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (therewasaguy) [reply src/amplitude.js:1]: hmm, that's a good point! Still, we don't need strict mode in our modules (any files in the `src` directory) because they're already in strict mode even after compilation with es6.

We can verify that by removing strict mode, and then doing any of the things that aren't allowed in es5's strict mode from this article https://flaviocopes.com/javascript-strict-mode/  and verifying that it still throws an error for the developer. My understanding of strict mode is that its audience is primarily developers because it throws errors that would otherwise be silent in ES5. So people who are working on p5.sound library will still get the same errors now that we're using ES6 modules. Meanwhile, people using the compiled library won't be modifying the library, so they don't need the library to be in strict mode.

We do still want to have `'use strict';` in in test files, since we haven't turned those into es6 modules yet.

File: lib/p5.sound.min.js [hunk 10]
```diff
@@ -1,3 +1,3 @@
...
...,o=Math.PI/180,r=0;r<44100;++r)e=2*r/44100-1,i[r]=(3+n)*e*20*o/(Math.PI+n*Math.abs(e));return i}p5.Distortion=function(t,e){if(i.call(this),void 0===t&&(t=.25),"number"!=typeof t)throw new Error("amount must be a number");if(void 0===e&&(e="2x"),"string"!=typeof e)throw new Error("oversample must be a String");var n=p5.prototype.map(t,0,1,0,2e3);this.waveShaperNode=this.ac.createWaveShaper(),this.amount=n,this.waveShaperNode.curve=o(n),this.waveShaperNode.oversample=e,this.input.connect(this.waveShaperNode),this.waveShaperNode.connect(this.wet)},p5.Distortion.prototype=Object.create(i.prototype),p5.Distortion.prototype.process=function(t,e,n){t.connect(this.input),this.set(e,n)},p5.Distortion.prototype.set=function(t,e){if(t){var n=p5.prototype.map(t,0,1,0,2e3);this.amount=n,this.waveShaperNode.curve=o(n)}e&&(this.waveShaperNode.oversample=e)},p5.Distortion.prototype.getAmount=function(){return this.amount},p5.Distortion.prototype.getOversample=function(){return this.waveShaperNode.oversample},p5.Distortion.prototype.dispose=function(){i.prototype.dispose.apply(this),this.waveShaperNode&&(this.waveShaperNode.disconnect(),this.waveShaperNode=null)}}.call(e,n,e,t))||(t.exports=i)}]);
```
Comment (therewasaguy) [inline lib/p5.sound.min.js:2]: The compiled versions of the library are still missing a lot of code, any idea why that is?

In general I'd recommend excluding `lib` files from PR's until we're ready to bump the library version number. But, if you want to include, make sure they compiled successfully !

File: src/amplitude.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (endurance21) [reply src/amplitude.js:1]: okay ! that's the nice  point to consider and i will be removing the 'strict mode'

**Comment by endurance21** [comment]
> it looks like `oscillator` has not been updated, and it's throwing some errors when testing locally.
> 
> also, be sure to either update the compiled library, or exclude it from the PR/commits

added the oscillator updates !

File: lib/p5.sound.min.js (hunk 10 above)
Comment (endurance21) [reply lib/p5.sound.min.js:2]: @therewasaguy  
i am not focusing on lib/ in my latest commits , and if you are checking it after my latest commit , build the file once again , i dont think the file is shrunk , it is 443kb as seen on linux

File: src/amplitude.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (endurance21) [reply src/amplitude.js]: sucessfully removed the 'strict mode' declaration!

**Comment by endurance21** [comment]
@therewasaguy  i have not pushed the latest build file so , while you are running tests locally using my latest commit , make sure to build it once !

File: src/effect.js [hunk 11]
```diff
@@ -1,165 +1,163 @@
...
+  if (this.wet) {
+    this.wet.disconnect();
+    delete this.wet;
+  }
+
+  this.ac = undefined;
+};
+
+export default p5.Effect;
+//why can't we do export p5.Effect ?
```
Comment (therewasaguy) [inline src/effect.js:163]: great question @endurance21 ! Without the default keyword, we'd be using a named export, and I guess "p5.Effect" is not a valid name.

We could do something like
`export const Effect = p5.Effect;` because then we'd be giving the export a valid name, which we could import as `import { Effect } from 'effect'`

and let's remove this comment from the code :)

File: src/oscillator.js [hunk 12]
```diff
@@ -1,634 +1,630 @@
...
+
+  // set default output gain to 0.5
+  this.output.gain.value = 0.5;
+  this.output.gain.setValueAtTime(0.5, p5sound.audiocontext.currentTime);
+
+  this.oscillator.connect(this.output);
+  // stereo panning
+  this.panPosition = 0.0;
+  this.connection = p5sound.input; // connect to p5sound by default
+  this.panner = new p5.Panner(this.output, this.connection, 1);
```
Comment (therewasaguy) [inline src/oscillator.js:105]: just an observation: there are a few examples like this where we're dependent on the order of imports. If app.js imported oscillator before panner, `p5.Panner` would not exist.

A better approach would be to `import p5Panner from 'panner'` here in the oscillator module.

This doesn't need to be done as part of this PR, but I noticed we're not exporting anything from p5.Panner right now...should we be consistent and have at least a default export for every module?

File: src/effect.js (hunk 11 above)
Comment (endurance21) [reply src/effect.js:163]: we can also do smth like this 
`
const Effect  = p5.Effect
export default Effect 
`

and we dont have to chnage anything in other files importing them

File: src/oscillator.js (hunk 12 above)
Comment (endurance21) [reply src/oscillator.js:105]: we ll see if we need to export smth out of  them in future !

File: src/oscillator.js [hunk 13]
```diff
@@ -627,4 +626,4 @@ p5.SqrOsc = function (freq) {
 
 p5.SqrOsc.prototype = Object.create(p5.Oscillator.prototype);
 
-export default p5.Oscillator;
```
Comment (therewasaguy) [inline src/oscillator.js]: why this change?

Not opposed to it but I do think we should be consistent about using names exports (like this one) or default exports?

**Comment by endurance21** [comment]
@therewasaguy  also there was p5. Js file added in that folder only for testing purpose,

So finally i think i have to do is

1.remove the p5. Js file from lib
2.build using grunt and commit those build file
??

**Comment by therewasaguy** [comment]
yes, that sounds good @endurance21

**Comment by endurance21** [comment]
@therewasaguy i  have added build files , take a look at that !

**Comment by therewasaguy** [comment]
@endurance21 feel free to "squash-and-merge"
